File: Olympiad/COCI/official/2009/contest1_solutions/jez.pas
program jez;

const LOG_MAXN = 20;

var
   R, S, nd, visak  : longint;
   red, najdulja, i : longint;
   K, ukupno, rez   : int64;

function min(a, b : longint) : longint; begin if a <= b then min := a else min := b; end;
function max(a, b : longint) : longint; begin if a >= b then max := a else max := b; end;

function popcount(x : longint):longint;
var rez : longint;
begin
   rez := 0;
   while x <> 0 do begin
      inc(rez);
      x := x - (x and -x);
   end;
   popcount := rez;
end;

function prebroji(x, ymax : longint):longint;
var k, jed, rez : longint;
begin
   rez := 0;
   for k:=LOG_MAXN downto 0 do begin
      if (ymax shr k) and 1 <> 0 then begin
         (* jed = broj jedinica u x na pozicijama 0..k-1 *)
         jed := popcount(x and ((1 shl k) - 1));
         rez := rez + (1 shl (k - jed));

         (* ako x ima jedinicu na poziciji k, svi brojevi izmedu 2^k i
            ymax imaju neki zajednicki bit, pa smo gotovi *)
         if (x shr k) and 1 <> 0 then break;
      end;
   end;
   prebroji := rez;
end;

function jedanred(red : longint) : longint;
var smax, sr : longint;
begin
   smax := nd-red;
   if nd mod 2 = 0 then begin
      sr := nd - max(0, nd-R+1);
      if (sr-red >= 0) and (sr-red < visak) then inc(smax);
   end else begin
      sr := max(0, nd-S+1);
      if (red-sr >= 0) and (red-sr < visak) then inc(smax);
   end;

   smax := max(smax, 0);
   smax := min(smax, S);
   jedanred := prebroji(red, smax);
end;

function dodaj(var ukupno : int64; x : longint) : boolean;
begin
   ukupno := ukupno + x;
   if ukupno > K then begin
      dodaj := true;
      exit;
   end;
   inc(nd);
   visak := K - ukupno;
   dodaj := false;
end;

begin
   read(R, S, K);
   najdulja := min(R, S);
   ukupno := 0;
   nd := 0;
   for i:=1 to najdulja do
      if dodaj(ukupno, i) then break;
   for i:=0 to R+S-2*najdulja-1 do
      if dodaj(ukupno, najdulja) then break;
   for i:=najdulja-1 downto 1 do 
      if dodaj(ukupno, i) then break;

   rez := 0;
   for red:=0 to R-1 do
      rez := rez+jedanred(red);
   writeln(rez);
end.

File: Olympiad/COCI/official/2009/contest1_solutions/jez.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>
using namespace std;

const int LOG_MAXN = 20;

int R, S; long long K;
int nd, visak;

int prebroji(int x, int ymax) {
   int rez = 0;
   for ( int k=LOG_MAXN; k>=0; --k ) {
      if ( (ymax&(1<<k)) != 0 ) {
         // jed = broj jedinica koje x ima na pozicijama 0..k-1
         int jed = __builtin_popcount(x&((1<<k)-1));
         rez += 1<<(k-jed);

         // ako x ima jedinicu na poziciji k, svi brojevi izmedu 2^k i
         // ymax imaju neki zajednicki bit, pa smo gotovi
         if ( (x&(1<<k)) != 0 ) break;
      }
   }
   return rez;
}

int jedanred(int r) {
   int smax = nd-r;
   if ( nd%2 == 0 ) {
      int sr = nd - max(0, nd-R+1);
      smax += sr-r >= 0 && sr-r < visak;
   } else {
      int sr = max(0, nd-S+1);
      smax += r-sr >= 0 && r-sr < visak;
   }
   smax = max(smax, 0);
   smax = min(smax, S);
   return prebroji(r, smax);
}

void dodaj(long long &ukupno, int x) {
   ukupno += x;
   if ( ukupno > K ) throw 1;
   ++nd;
   visak = K - ukupno;
}

int main() {
   cin >> R >> S >> K;

   try {
      int najdulja = min(R, S);
      long long ukupno = 0;
      nd = 0;
      for ( int i=1; i<=najdulja; ++i ) dodaj(ukupno, i);
      for ( int i=0; i<R+S-2*najdulja; ++i ) dodaj(ukupno, najdulja);
      for ( int i=najdulja-1; i>0; --i ) dodaj(ukupno, i);
   } catch(...) { }

   long long rez = 0;
   for ( int r=0; r<R; ++r )
      rez += jedanred(r);
   cout << rez << endl;

   return 0;
}
